net/nettest/pipe.go

```diff
@@ -145,117 +148,94 @@ func (p *Pipe) Close() error {
 		p.cancelReadTimer()
 		p.cancelReadTimer = nil
 	}
-	p.mu.Unlock()
+	p.cnd.L.Unlock()
 
 	if closed {
-		return fmt.Errorf("nettest.Pipe(%q).Close: already closed", p.name)
+		return net.ErrClosed
 	}
+	p.cnd.Broadcast()
 
-	p.signalRead()
-	p.signalWrite()
 	return nil
 }
 
+func (p *Pipe) deadlineTimer(t time.Time) func() {
+	if t.IsZero() {
+		return nil
+	}
+	d := time.Until(t)
+	if d < 0 {
+		p.cnd.Broadcast()
+		return nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
use https://golang.org/pkg/context/#WithDeadline instead and ditch time.Until and time.NewTimer?